service/batchmdw/batch_processor.go

```diff
@@ -0,0 +1,142 @@
+package batchmdw
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"sync"
+
+	"github.com/kava-labs/kava-proxy-service/service/cachemdw"
+)
+
+// BatchProcessor makes multiple requests to the underlying handler and then combines all the
+// responses into a single response.
+// It assumes all individual responses are valid json. Each response is then marshaled into an array.
+type BatchProcessor struct {
+	handler   http.HandlerFunc
+	requests  []*http.Request
+	responses []*bytes.Buffer
+	header    http.Header
+	cacheHits int
+	status    int
+	mu        sync.Mutex
+}
+
+// NewBatchProcessor creates a BatchProcessor for combining the responses of reqs to the handler
+func NewBatchProcessor(handler http.HandlerFunc, reqs []*http.Request) *BatchProcessor {
+	return &BatchProcessor{
+		handler:   handler,
+		requests:  reqs,
+		responses: make([]*bytes.Buffer, len(reqs)),
+		header:    nil,
+		status:    http.StatusOK,
+		mu:        sync.Mutex{},
+	}
+}
+
+// RequestAndServe concurrently sends each request to the underlying handler
+// Responses are then collated into a JSON array and written to the ResponseWriter
+func (bp *BatchProcessor) RequestAndServe(w http.ResponseWriter) error {
+	wg := sync.WaitGroup{}
+	for i, r := range bp.requests {
```

Comment:
unlikely but potential DDOS vector, might make sense to pick some reasonably high limit (100, 1000?) configured via env for how much request amplification a single batched request can place on the backend fleet

Reply:
great point, didn't think through the fact that this removes _any_ batch limit that may be imposed by the actual nodes 😅 🙏

Reply:
i updated to have the batch processing middleware take and respect a maximum batch size. if it gets a request more sub-requests than that, it responds 413.